### DISCUSSION TITLE
Send notifications to an API based notification service ?

### DISCUSSION THREAD

**Discussion by KingBain** [luc-github/ESP3D#704] [Questions]
Is it possible to send notification via a HTTP post ?

Its not obvious to me in the documentation, but I see so many other discussions about integration with Alexa and MQTT that I feel its possible ?

**Comment by KingBain** [comment]
It looks like it is possible, but maybe requires a modification to the code. 

looks like the LINE service is using the LINE API 
https://github.com/luc-github/ESP3D/blob/ecf48fe89de215052c388a29eb28556bb483911b/esp3d/notifications_service.cpp#L297

**Comment by luc-github** [comment]
you can send notification using http using the [ESP600] command and the webcommand handler

`http://localhost:8088/command?cmd=%5BESP600%5DThis+is+message` put your ESP3D Ip instead of `localhost:8088` of course

**Reply by KingBain** [reply]
thanks ill check that out, probably peek at the code too. some of these api services look for special headers or json payloads

**Reply by KingBain** [reply]
Hey sorry, I think there might be some confusion. 

I don't want to send a comment to the printer via the webservice run by ESP3D, I want the printer to send a message to an external webhook or other API service. 

### As an example 

I have a line at the end of the GCODE file that looks like this, _sorry pseudo code_ 

### Last Line
`<send_web_request> <to_url https://maker.ifttt.com/trigger/printer_notify/with/key/ovL4asdfasdfasdfUmdwBOarilxoxQF-OcVncL-jeYH9mZezax0au3X?value1=3D Printer54&value2=Completed>  `


_Maker.ifttt.com_ is the incoming webhook service of IFTTT and I'm passing 4 variables in the URL that Maker is looking for :

1. Maker event "printer_notify"
2. webhook authentication key  "ovL4asdfasdfasdfUmdwBOarilxoxQF-OcVncL-jeYH9mZezax0au3X"
3. value1 = "3D Printer54"
4. value2 = "Completed"

In IFTTT, the webhook event looks like 
![image](https://user-images.githubusercontent.com/367922/148496687-0645f9b3-7348-4e33-9447-afe61c0a14dd.png)

It's basically one incoming event and one outgoing event. 

### Incoming event 
![image](https://user-images.githubusercontent.com/367922/148496799-99bf82d6-512e-4adb-9acb-673a8cdf7e05.png)

### Outgoing event

![image](https://user-images.githubusercontent.com/367922/148497300-34b61b7b-e21e-4a02-9dd8-b958daae7a37.png)
![image](https://user-images.githubusercontent.com/367922/148497474-e26efa09-3abe-437f-a85a-c9c34477de36.png)

So I want it to go from the printer, to IFTTT, to another API notification service

I'm using IFTTT, because the API service that I really want to use only accepts JSON payloads and needs an authentication header.

I have most of this working URL process-wise, just not sure how to make ESP3D fire off the URL. 

If I use a tool like Postman or curl, I can see the events fire properly.

POST to https://maker.ifttt.com/trigger/printer_notify/with/key/ovL4asdfasdfasdfUmdwBOarilxoxQF-OcVncL-jeYH9mZezax0au3X?value1=3D Printer54&value2=Completed

IFTTT grabs it
![image](https://user-images.githubusercontent.com/367922/148497789-19a4a933-4f81-4d77-a9ed-a33d5a06af11.png)

IFTTT then fire off its own request 
![image](https://user-images.githubusercontent.com/367922/148497913-bbbb0777-ee6b-4c03-b3cd-3097fb5c3b67.png)

The government of Canada API service then catches that incoming web request and generate a text message and an email that gets sent to my phone 

which lookalike this
![image](https://user-images.githubusercontent.com/367922/148498173-c409ea13-5f55-4185-a8e5-962d7376f694.png)

**Reply by luc-github** [reply]
here what is supported : https://github.com/luc-github/ESP3D/wiki/Notifications
There is no `generic` API - all will depend of service provided, the only generic is the email

**Comment by sineva54** [comment]
Hi, how can I send messages with filename & ETA time?! Thanks!
p/s: My Printer use Marlin & ESP3D uses LINE Notify

**Reply by luc-github** [reply]
there is no variable for eta / filename you need to put them manually or add them using post processing script in your slicer

**Reply by sineva54** [reply]
What about Runout Filament, STOP, Abort...!? Can ESP3D detect these events and send messages !?!

**Reply by luc-github** [reply]
Marlin has  configurable script where you can add notifications commands, e.g: M600 for filament change when filament sensor is supported
but ESP3d it self has no way to detect printer event unless printer inform esp3d

**Reply by sineva54** [reply]
Thanks! I'll try it
